extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchBufferAggregator.java

```diff
@@ -87,12 +81,27 @@ public Object get(ByteBuffer buf, int position)
   //Note that this is not threadsafe and I don't think it needs to be
   private Union getUnion(ByteBuffer buf, int position)
   {
-    Union union = unions.get(position);
-    if (union == null) {
-      Memory mem = new MemoryRegion(nm, position, maxIntermediateSize);
-      union = (Union) SetOperation.wrap(mem);
-      unions.put(position, union);
+    Int2ObjectMap<Union> unionMap = unions.get(buf);
+    Union union = unionMap != null ? unionMap.get(position) : null;
+    if (union != null) {
+      return union;
+    }
+    return createNewUnion(buf, position, true);
+  }
+
+  private Union createNewUnion(ByteBuffer buf, int position, boolean isWrapped)
+  {
+    NativeMemory nm = getNativeMemory(buf);
+    Memory mem = new MemoryRegion(nm, position, maxIntermediateSize);
+    Union union = isWrapped
+                  ? (Union) SetOperation.wrap(mem)
+                  : (Union) SetOperation.builder().initMemory(mem).build(size, Family.UNION);
+    Int2ObjectMap<Union> unionMap = unions.get(buf);
```

Comment:
> Then you can make it much more faster (compared to hand-written/computeIfAbsent difference) if you get rid of those IdentitiyHashMaps altogether. Your objections here: #4071 (comment) are for possible-future-BufferAggregator-users who violate the contract of calling aggregate() with the same ByteBuffer (serially).

Is providing different ByteBuffer objects really violating the contract of BufferAggregator? I can't read anything in the javadocs that suggests this. I guess this matters to me because I wrote the code that is making this change necessary (when BufferGrouper resizes itself, it might move aggregator states into a new buffer) and at the time I wrote it, I didn't think I was violating any contract.

> Or keep IdentityHashMap, but add fields cachedBB and cachedUnionMap and add something like if (bb == cachedBB) return cachedUnionMap.get(position) in the beginning of getUnion().

This would work, and in fact I thought about suggesting it in a review, but then refrained since I suspect that even though `getUnion` is called as part of the hot `aggregate` method, `SketchAggregator.updateUnion` probably takes a lot more time anyway.